File: android/app/src/main/kotlin/dev/tvolkert/codeword/NetworkChannel.java
package dev.tvolkert.codeword;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.LinkAddress;
import android.net.LinkProperties;
import android.net.Network;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class NetworkChannel {
    private static final String TAG = "flutter";
    private static final String CHANNEL_NAME = "codeword.tvolkert.dev/network";
    private static final String METHOD_NAME_GET_ADDRESSES = "getAddresses";
    public static final String ERROR_CODE_NO_NETWORK = "no_network";
    public static final String ERROR_CODE_INCOMPATIBLE_DEVICE = "incompatible_device";

    @NonNull private final FlutterEngine flutterEngine;
    @NonNull private final ConnectivityManager connectivityManager;

    @Nullable private MethodChannel channel;

    NetworkChannel(@NonNull FlutterEngine flutterEngine, @NonNull Context applicationContext) {
        this.flutterEngine = flutterEngine;
        this.connectivityManager = (ConnectivityManager) applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    public void register() {
        if (channel != null) {
            // Already registered.
            return;
        }

        channel = new MethodChannel(flutterEngine.getDartExecutor().getBinaryMessenger(), CHANNEL_NAME);
        channel.setMethodCallHandler((MethodCall methodCall, MethodChannel.Result result) -> {
            switch (methodCall.method) {
                case METHOD_NAME_GET_ADDRESSES:
                    getAddresses(result);
                    break;
                default:
                    result.notImplemented();
            }
        });
    }

    private void getAddresses(MethodChannel.Result result) {
        List<String> data = new ArrayList<String>();
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            @Nullable final Network network = connectivityManager.getActiveNetwork();
            if (network == null) {
                result.error(ERROR_CODE_NO_NETWORK, null, null);
            }

            @Nullable final LinkProperties linkProperties = connectivityManager.getLinkProperties(network);
            if (linkProperties == null) {
                result.error(ERROR_CODE_NO_NETWORK, null, null);
            }

            assert linkProperties != null;
            final List<LinkAddress> linkAddresses = linkProperties.getLinkAddresses();
            for (LinkAddress linkAddress : linkAddresses) {
                if (linkAddress.getAddress().getAddress().length == 4) {
                    // IPv4 address.
                    final String address = linkAddress.getAddress().getHostAddress();
                    final int prefixLength = linkAddress.getPrefixLength();
                    data.add(address + "/" + prefixLength);
                }
            }
        } else {
            result.error(ERROR_CODE_INCOMPATIBLE_DEVICE, null, null);
        }
        result.success(data);
    }
}

File: android/app/src/main/kotlin/dev/tvolkert/codeword/NetworkChannel.kt
package dev.tvolkert.codeword

import android.content.Context
import android.net.ConnectivityManager
import android.net.LinkProperties
import android.net.Network

import java.util.ArrayList
import kotlin.collections.MutableList

import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class NetworkChannel internal constructor(
    private val flutterEngine: FlutterEngine,
    applicationContext: Context
) {
    private val connectivityManager: ConnectivityManager =
        applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    private var channel: MethodChannel? = null

    fun register() {
        if (channel != null) {
            // Already registered.
            return
        }

        val localChannel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL_NAME)
        channel = localChannel
        localChannel.setMethodCallHandler { methodCall: MethodCall, result: MethodChannel.Result ->
            when (methodCall.method) {
                METHOD_NAME_GET_ADDRESSES -> getAddresses(result)
                else -> result.notImplemented()
            }
        }
    }

    private fun getAddresses(result: MethodChannel.Result) {
        val data: MutableList<String> = ArrayList<String>()
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            val network: Network? = connectivityManager.activeNetwork
            if (network == null) {
                result.error(ERROR_CODE_NO_NETWORK, null, null)
            }

            val linkProperties: LinkProperties? = connectivityManager.getLinkProperties(network)
            if (linkProperties == null) {
                result.error(ERROR_CODE_NO_NETWORK, null, null)
            }

            checkNotNull(linkProperties)
            for (linkAddress in linkProperties.linkAddresses) {
                if (linkAddress.address.address.size == 4) {
                    // IPv4 address.
                    val address: String = linkAddress.address.hostAddress!!
                    val prefixLength: Int = linkAddress.prefixLength
                    data.add("$address/$prefixLength")
                }
            }
        } else {
            result.error(ERROR_CODE_INCOMPATIBLE_DEVICE, null, null)
        }
        result.success(data)
    }

    companion object {
        private const val CHANNEL_NAME = "codeword.tvolkert.dev/network"
        private const val METHOD_NAME_GET_ADDRESSES = "getAddresses"
        const val ERROR_CODE_NO_NETWORK: String = "no_network"
        const val ERROR_CODE_INCOMPATIBLE_DEVICE: String = "incompatible_device"
    }
}
